### ISSUE TITLE
Missing feature to reproduce SoundStream's Residual Vector Quantizer

### ISSUE THREAD

**Issue by wesbz** [lucidrains/vector-quantize-pytorch#3]
Hi,
Thanks for this cool work!
I couldn't help but notice that a few features used to improve the usage of the codebooks were missing to be an exact implementation of the work done in the [SoundStream](https://arxiv.org/pdf/2107.03312.pdf) article.
- _"First, instead of using a random initialization for the codebook vectors, we run the k-means algorithm on the first training batch and use the learned centroids as initialization"_
- _"Second, as proposed in [34], when a codebook vector has not been assigned any input frame for several batches, we replace it with an input frame randomly sampled within the current batch."_
I'm currently working on an implementation of this work, I'll use and adapt your code for this purpose but thought you might want to know about it.
I'll keep you posted :)

**Comment by wesbz** [comment]
@lucidrains thanks for your response!
I have completed the first bullet point, but I'd be more than glad to see your implementation of this!

**Comment by wesbz** [comment]
I completed (I think) both points, you can see it [here](https://github.com/wesbz/SoundStream)

**Comment by lucidrains** [comment]
@wesbz also, would you be interested in testing https://github.com/lucidrains/vector-quantize-pytorch/issues/2#issuecomment-937387129 if i were to build it?

**Comment by lucidrains** [comment]
@wesbz added my version of bullet point 2 here https://github.com/lucidrains/vector-quantize-pytorch#expiring-stale-codes

**Comment by lucidrains** [comment]
@wesbz ok sounds good! thank you for the PR! i realized there's an issue with the cluster size; we should probably be keeping track of the init state across all codes, and on the first EMA, just set the cluster size to the value instead of assuming 0 for the moving average

I'll check out the dropout later this evening as well!
